src/sentry/interfaces/stacktrace.py

```diff
@@ -837,48 +931,61 @@ def to_json(self):
-            return []
-
+    def get_grouping_component(self, platform=None, variant=None):
         frames = self.frames
+        contributes = None
+        hint = None
+        all_frames_are_in_app = False
 
         # TODO(dcramer): this should apply only to platform=javascript
         # Browser JS will often throw errors (from inlined code in an HTML page)
         # which contain only a single frame, no function name, and have the HTML
         # document as the filename. In this case the hash is often not usable as
         # the context cannot be trusted and the URL is dynamic (this also means
         # the line number cannot be trusted).
-        stack_invalid = (len(frames) == 1 and not frames[0].function and frames[0].is_url())
-
-        if stack_invalid:
-            return []
-
-        system_frames = variant == 'system'
-        if not system_frames:
+        if (len(frames) == 1 and not frames[0].function and frames[0].is_url()):
+            contributes = False
+            hint = 'ignored single frame stack'
+        elif variant == 'app':
             total_frames = len(frames)
-            frames = [f for f in frames if f.in_app] or frames
+            in_app_count = sum(1 if f.in_app else 0 for f in frames)
+            if in_app_count == 0:
+                in_app_count = total_frames
+                all_frames_are_in_app = True
 
             # if app frames make up less than 10% of the stacktrace discard
             # the hash as invalid
-            if len(frames) / float(total_frames) < 0.10:
-                return []
-
-        if not frames:
-            return []
-
-        output = []
-
-        # stacktraces that only differ by the number of recursive calls should
-        # hash the same, so we squash recursive calls by comparing each frame
-        # to the previous frame
-        output.extend(frames[0].get_hash(platform, variant=variant))
-        prev_frame = frames[0]
-        for frame in frames[1:]:
-            if not is_recursion(frame, prev_frame):
-                output.extend(frame.get_hash(platform, variant=variant))
+            if total_frames > 0 and in_app_count / float(total_frames) < 0.10:
+                contributes = False
+                hint = 'less than 10% of frames are in-app'
+
+        values = []
+        prev_frame = None
+        for frame in frames:
+            frame_component = frame.get_grouping_component(platform, variant)
+            if variant == 'app' and not frame.in_app and not all_frames_are_in_app:
+                frame_component.update(
+                    contributes=False,
+                    hint='non app frame',
+                )
+            elif prev_frame is not None and is_recursion(frame, prev_frame):
+                frame_component.update(
+                    contributes=False,
+                    hint='ignored due to recursion',
+                )
+            elif variant == 'app' and not frame.in_app and all_frames_are_in_app:
```

Comment:
Kinda tricky to read this because it's not obvious the recursion check needs to be performed first.